#2947: fix for wildcard bounded values.

Zegveld opened the pull request:
Got the given example situation working and all other tests still pass.
- [ ] Add testcase with super wildcard.
- [ ] rename test package to be more generic for extends/super wildcard usage. maybe `typebound` instead of `wildcardextends`.

Already created the PR so that the fix is already reviewable. (Only planning on adding more tests not changing more code)

filiphr commented:
I also had a look at this one when I was analysing it and it was tricky, because on the first step like you've done now it is being fixed. However, I think that if you do `@Mapping( target = "object", source = "object.contents" )` it will no longer work, or something similar to that. I think once you do without bounds for `<?>` you get `Object` type for `?`. However, it is worth exploring ways to fix this properly.

filiphr commented:
@Zegveld I really dislike the fix in this PR. It is way to intrusive and changes a lot of things that shouldn't really be affected. I'll give it a go and try to see if I can come up with another fix for this one.

Zegveld commented:
The one bit I don't like myself about the current change is the "?" comparison bit at the `Filters` class.
Though the changes at `Type` feel logical to me.

As a result of the changes an parameter at `TypeFactory.getReturnType(...)` became obsolete which resulted in the changes there.

filiphr commented:
>Though the changes at `Type` feel logical to me.

All that looks way too error prone. We should create a correct accessor with a correct type and not pass this things down to the `Type` to resolve it. When `@Mapping` is not used everything works fine, so we need to only fix it in the creation of the `SourceReference`. Perhaps we need to tackle #2070 once and for all :)

Zegveld commented:
Seeing as tackling #2070 would indeed mean that the code there would change again, I've done some more rework and moved all the required actions onto the `Filters` class. At the same time I've refactored the "?" comparison away. Still not 100% happy with the new result but it is improving.

sjaakd commented:
I guess I need to dig in to this material as well. I'll try to have a look sometime this week..

Zegveld commented:
I believe the current state is the best I can do atm. if you have any idea's for improvements I'm open to them.

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/util/Filters.java:
```diff
@@ -111,7 +114,17 @@ public Map<String, ReadAccessor> recordAccessorsIn(Collection<Element> recordCom
     private TypeMirror getReturnType(ExecutableElement executableElement) {
-        return getWithinContext( executableElement ).getReturnType();
+        TypeMirror returnType = getWithinContext( executableElement ).getReturnType();
+        if ( returnType.getKind() == TypeKind.WILDCARD ) {
```
hmm. can return type be wildcards (that have an extends bound)?

sjaakd reviewed (commented):
Its a while ago.. but for me this: http://www.angelikalanger.com/GenericsFAQ/JavaGenericsFAQ.html was really useful. I'll have a better look at the case soon.

filiphr commented:
I really do not feel comfortable removing one of the methods from the `TypeFactory`. Imagine if we have something like:

```java
public interface OtherWildCardInterface extends WildcardInterface {
}
```

```java
@Mapper
public interface WildcardNestedInheritedExtendsMapper {
    WildcardNestedInheritedExtendsMapper INSTANCE = Mappers.getMapper( WildcardNestedInheritedExtendsMapper.class );

    @Mapping( target = "object", source = "contained.value")
    Target map(SourceContainerInherited<? extends OtherWildCardInterface> source);

    default String mapToString(WildcardInterface rawData) {
        return rawData.getContents();
    }

    default String mapToString(OtherWildCardInterface rawData) {
        return rawData.getContents() + "other";
    }
}
```

Or perhaps another raw types that need to be returned in the scope of the currently declared type. There might be some other edge cases where we resolve these things.

---

In addition to that in the `Filters` we have

```java
        if ( returnType.getKind() == TypeKind.WILDCARD ) {
            if ( ( (WildcardType) returnType ).getExtendsBound() != null ) {
                return ( (WildcardType) returnType ).getExtendsBound();
            }
            return ( (TypeVariable) executableElement.getReturnType() ).getUpperBound();
        }
```

I am confused by that. We check the `returnType` and if that does not have an extends bound we assume that we can cast the executable element to `TypeVariable`. 

---

I'll need some more time to process this changes and try to wrap my head around them

Zegveld commented:
with `SourceContainerInherited<? extends OtherWildCardInterface>` then the type returned is `OtherWildCardInterface`, because the returnType of the container internal method is a `WILDCARD` with an `extendsBound`, so that `extendsBound` is chosen.

I'll move the `Issue2677Mapper` and corresponding test to the location where the new tests are as well. Then this case is also visible in the code-review.

----

concerning the Filters bit `returnType.getKind() == TypeKind.WILDCARD` can only be true when `executable.getReturnType()` is a `TypeVariable`. 

It however is not true that because `executable.getReturnType()` is a `TypeVariable` then the `returnType.getKind()` must be a wildcard. There are other types that also fit the `TypeVariable` grouping, like `TypeKind.TYPEVAR`.

----

Think I'll also add some inline comments as to how and why the code works like this, and which situation is covered by what line.

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/util/Filters.java:
```diff
@@ -111,7 +114,24 @@ public Map<String, ReadAccessor> recordAccessorsIn(Collection<Element> recordCom
 
     private TypeMirror getReturnType(ExecutableElement executableElement) {
-        return getWithinContext( executableElement ).getReturnType();
+        TypeMirror returnType = getWithinContext( executableElement ).getReturnType();
```
rename `getWithinContext` to `getMethodAsMemberOfDefiningClass`. guess thats a more appropriate name

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/util/Filters.java:
```diff
@@ -111,7 +114,24 @@ public Map<String, ReadAccessor> recordAccessorsIn(Collection<Element> recordCom
+                // in case of '? extends ...' we want to return the '...' part.
+                return ( (WildcardType) returnType ).getExtendsBound();
+            }
+            // in case of just '?', we want to look at the original methods return type and see if that one has an
```
I guess this can only be if the method is generic. Because the `getWithinContext` would already have tackled the class defined type-var's right?

Zegveld replied:
That is correct.

sjaakd commented on processor/src/test/java/org/mapstruct/ap/test/generics/typebounds/WildcardedInterface.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * @author Ben Zegveld
+ */
+public interface WildcardedInterface {
```
I understand that this is the base for a wildcard.. But the interface itself does not contain wildcards. Hence the name is somewhat confusing. Would `UpperBound` be a [better name](http://www.angelikalanger.com/GenericsFAQ/FAQSections/TypeArguments.html#What%20is%20a%20bounded%20wildcard?)?

sjaakd replied:
Or `WildCardBase` or just `WildCard`

Zegveld replied:
went with `WildCard` and `WildCardImpl`

sjaakd commented on processor/src/test/java/org/mapstruct/ap/test/generics/typebounds/WildcardedInterfaceImpl.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * @author Ben Zegveld
+ */
+class WildcardedInterfaceImpl implements WildcardedInterface {
```
same here: `UpperBoundImpl`?

Zegveld replied:
went with `WildCard` and `WildCardImpl`

sjaakd reviewed (commented):
LGTM.. some naming remarks..
